File: go/elons-toys/elons_toys.go
package elon

import "fmt"

// the Drive method updates the number of metres driven based on
// a car's speed and reduces the battery accordingly
func (car *Car) Drive() {
	if car.battery < car.batteryDrain {
		return
	}
	car.distance += car.speed
	car.battery -= car.batteryDrain
}

// DisplayDistance returns the distance as displayed on the LED display
func (car Car) DisplayDistance() string {

	return fmt.Sprintf("Driven %d meters", car.distance)
}

// DisplayDistance returns the battery percentage as displayed on the LED display
func (car Car) DisplayBattery() string {

	return fmt.Sprintf("Battery at %d%%", car.battery)
}

// CanFinish takes a trackDistance int as its parameter and returns true
// if the car can finish the race; otherwise, returns false
func (car Car) CanFinish(trackDistance int) bool {
	maxTravelDistance := (car.battery / car.batteryDrain) * car.speed
	return maxTravelDistance >= trackDistance
}

File: csharp/elons-toys/ElonsToys.cs
using System;

class RemoteControlCar
{
    private int _distance = 0;
    private int _battery = 100;

    private int _speed = 20;
    private int _batteryDrain = 1;

    public static RemoteControlCar Buy()
    {
        return new RemoteControlCar();
    }

    public string DistanceDisplay()
    {
        return $"Driven {_distance} meters";
    }

    public string BatteryDisplay()
    {
        if (_battery == 0) return "Battery empty";
        return $"Battery at {_battery}%";
    }

    public void Drive()
    {
        if (_battery > 0)
        {
            _battery -= _batteryDrain;
            _distance += _speed;
        }
    }
}
